python/lsst/obs/base/fitsRawFormatterBase.py

```diff
@@ -118,21 +124,83 @@ def makeVisitInfo(self, metadata):
         """
         raise NotImplementedError("Must be implemented by subclasses.")
 
-    def makeWcs(self, metadata):
-        """Construct a SkyWcs from metadata.
+    @abstractmethod
+    def getDetector(self, id):
+        """Return the detector that acquired this raw exposure.
 
         Parameters
         ----------
+        id : `int`
+            The identifying number of the detector to get.
+
+        Returns
+        -------
+        detector : `~lsst.afw.cameraGeom.Detector`
+            The detector associated with that ``id``.
+        """
+        raise NotImplementedError("Must be implemented by subclasses.")
+
+    def makeWcs(self, visitInfo, detector, metadata):
+        """Create a SkyWcs from information about the exposure.
+
+        If VisitInfo is not None, use it and the detector to create a SkyWcs,
+        otherwise return the metadata-based SkyWcs (always created, so that
+        the relevant metadata keywords are stripped).
+
+        Parameters
+        ----------
+        visitInfo : `~lsst.afw.image.VisitInfo`
+            The information about the telescope boresight and camera
+            orientation angle for this exposure.
+        detector : `~lsst.afw.cameraGeom.Detector`
+            The detector used to acquire this exposure.
         metadata : `~lsst.daf.base.PropertyList`
             Header metadata.  May be modified in-place.
 
         Returns
         -------
-        wcs : `~lsst.afw.geom.SkyWcs`
+        skyWcs : `~lsst.afw.geom.SkyWcs`
             Reversible mapping from pixel coordinates to sky coordinates.
+
+        Raises
+        ------
+        InitialSkyWcsError
+            Raised if there is an error generating the SkyWcs, chained from the
+            lower-level exception if available.
+        """
+        skyWcs = self._createSkyWcsFromMetadata(metadata)
+
+        log = lsst.log.Log.getLogger("fitsRawFormatter")
+        if visitInfo is None:
+            msg = "No VisitInfo; cannot access boresight information. Defaulting to metadata-based SkyWcs."
+            log.warn(msg)
+            if skyWcs is None:
+                raise InitialSkyWcsError("Failed to create both metadata and boresight-based SkyWcs."
+                                         "See warnings in log messages for details.")
+            return skyWcs
+        skyWcs = createInitialSkyWcs(visitInfo, detector)
+
+        return skyWcs
+
+    def _createSkyWcsFromMetadata(self, metadata):
```

Comment:
I feel like I'm having a bad case of deja vu. Why does this need to exist both here and in the cameraMapper? It suggests to me it maybe shouldn't be a sneaky `_` function and instead be one thing accessible both from here and from cameraMapper.py.

Reply:
I'm making the behavior different between gen2 (`cameraMapper.py`) and gen3 (`fitsRawFormatterBase.py`): gen2 logs a debug message because even calib exposure go through that. We're trying to make gen3 louder about failures, so it's a `warn`. Also gen3 just returns the WCS, instead of setting it on the `Exposure` object (gen2).